Source/WebCore/platform/gstreamer/GStreamerQuirkBroadcomBase.cpp

```diff
@@ -0,0 +1,192 @@
+#define GST_CAT_DEFAULT webkit_broadcom_base_quirks_debug
+
+GStreamerQuirkBroadcomBase::GStreamerQuirkBroadcomBase()
+{
+    GST_DEBUG_CATEGORY_INIT(webkit_broadcom_base_quirks_debug, "webkitquirksbroadcombase", 0, "WebKit Broadcom Base Quirks");
+}
+
+bool GStreamerQuirkBroadcomBase::queryBufferingPercentage(MediaPlayerPrivateGStreamer* playerPrivate, const char*& elementName, GRefPtr<GstQuery>& query) const
+{
+    if (!isEnsuredOwnedState(playerPrivate))
+        return false;
+    GStreamerQuirkBroadcomBaseState& state = static_cast<GStreamerQuirkBroadcomBaseState&>(playerPrivate->quirkState());
+
+    if (playerPrivate->shouldDownload() || !state.m_queue2)
+        return false;
+    if (gst_element_query(state.m_queue2.get(), query.get())) {
+        elementName = "queue2";
+        return true;
+    }
+    return false;
+}
+
+int GStreamerQuirkBroadcomBase::correctBufferingPercentage(MediaPlayerPrivateGStreamer* playerPrivate, int originalBufferingPercentage, GstBufferingMode mode) const
+{
+    if (!isEnsuredOwnedState(playerPrivate))
+        return originalBufferingPercentage;
+
+    GStreamerQuirkBroadcomBaseState& state = static_cast<GStreamerQuirkBroadcomBaseState&>(playerPrivate->quirkState());
+
+    // The Nexus playpump buffers a lot of data. Let's add it as if it had been buffered by the GstQueue2
+    // (only when using in-memory buffering), so we get more realistic percentages.
+    if (mode != GST_BUFFERING_STREAM || !state.m_vidfilter)
+        return originalBufferingPercentage;
+
+    // The Nexus playpump buffers a lot of data. Let's add it as if it had been buffered by the GstQueue2
+    // (only when using in-memory buffering), so we get more realistic percentages.
+    int correctedBufferingPercentage1 = originalBufferingPercentage;
+    int correctedBufferingPercentage2 = originalBufferingPercentage;
+    unsigned maxSizeBytes = 0;
+
+    // We don't trust the buffering percentage when it's 0, better rely on current-level-bytes and compute a new buffer level accordingly.
+    g_object_get(state.m_queue2.get(), "max-size-bytes", &maxSizeBytes, nullptr);
+    if (!originalBufferingPercentage && state.m_queue2) {
+        unsigned currentLevelBytes = 0;
+        g_object_get(state.m_queue2.get(), "current-level-bytes", &currentLevelBytes, nullptr);
+        correctedBufferingPercentage1 = currentLevelBytes > maxSizeBytes ? 100 : static_cast<int>(currentLevelBytes * 100 / maxSizeBytes);
+    }
+
+    unsigned playpumpBufferedBytes = 0;
+    if (state.m_vidfilter)
+        g_object_get(GST_OBJECT(state.m_vidfilter.get()), "buffered-bytes", &playpumpBufferedBytes, nullptr);
+
+    unsigned multiqueueBufferedBytes = 0;
+    if (state.m_multiqueue) {
+        GUniqueOutPtr<GstStructure> stats;
+        g_object_get(state.m_multiqueue.get(), "stats", &stats.outPtr(), nullptr);
+        const GValue* queues = gst_structure_get_value(stats.get(), "queues");
+        unsigned size = gst_value_array_get_size(queues);
```

Comment:
sorry, there is already unsigned, you don't need to define it, so you can use it already.

Reply:
In this very specific case (and only in this) gstStructureGet() doesn't make sense. That function is useful to avoid defining an extra variable to be passed to gst_structure_get_whatevertype() by reference, but in this case we're getting a GValue that holds a GArray. Those aren't supported by gstStructureGet() and addition of generic `const GValue*` values doesn't make sense, because those are already returned by gst_structure_get_value() as the result of the function call. They don't need to be passed as arguments by reference.

Therefore, I would just be changing one call by another to get the same type in the end.

However, using gstStructureGet() in the lines below to get basic types **does** make sense.